### ISSUE TITLE
"This statement seems to be disconnected from its parent foreach."

### ISSUE THREAD

**Issue by bitwombat** [kalessil/phpinspectionsea#173]
I have this in a phtml file:

```
            <?php
            if (count($currencies) > 0) {
                echo '<ul>';
                foreach ($currencies as $currency) {
                    $competency = $currency->getCompetency();
                    echo '<li>' . $competency->getName();
                    echo '   (exp. ' . $currency->getExpiryDate()->format('Y-m-d') . ')';
                    echo '</li>';    // **** phpinspectionsea doesn't like this
                }
                echo '</ul>';   
            } else {
                echo '<p>&lt;none&gt;</p>';
            }
            ?>
```

At the line indicated above, phpinspectionsea 2.3.4 indicates
"This statement seems to be disconnected from its parent foreach."

**Comment by kalessil** [comment]
Thank you for reporting @bitwombat, nice catch.

**Comment by funivan** [comment]
Other false-positive case
```php
$files = [
    '1.html',
    '2.html',
  ];
  foreach ($files as $index => $file) {
    ob_start(); // Error goes here
    include __DIR__ . '/' . $file;
    $map[$index] = ob_get_contents();
    ob_end_clean(); // Error goes here
  }
```

**Comment by kalessil** [comment]
For myself: statements which doesn;t reference variables should not be reported

**Comment by rentalhost** [comment]
I don't know if I understand your last message, but you should take care about things like:

```php
foreach ($messages as $message) {
    printf('hello');
}
```

It do not references any variable, _but should be reported_.

**Comment by funivan** [comment]
@rentalhost  should be not reported =)
If we have 10 messages then we get 10 hello, and this is correct.

**Comment by rentalhost** [comment]
I don't see reason for it allows things like that. Maybe if it is a `for()`, but on `foreach()` it doesn't make sense for me.

**Comment by kalessil** [comment]
I think we should cut a number of false-positives with this approach. 

In fact, we'll not lose much - real problems will be spotted.

**Comment by rentalhost** [comment]
Right, but if you cut cases like that, what will be really reported?

**Comment by kalessil** [comment]
E.g. in the case of Symfony (2.7 branch): from 19 cases, 17 will remain.
That's right not relying on just 1 project stats, but the approach gives us a chance to focus on relly weird staff.

**Comment by rentalhost** [comment]
Great! 

I guess that futurely with the implementation of #221 we can work on this topic again to improve this inspection (for instance, if called function have no sense inside of a looping).

**Comment by kalessil** [comment]
Yep =)

**Comment by TrueXakeP** [comment]
Is something wrong with this?
```php
$command = $this->db->createCommand('DELETE ls, lt FROM language_source ls INNER JOIN language_translate lt ON ls.id = lt.id WHERE ls.category = :category AND ls.message = :message');
$command->prepare(false);
foreach ($this->translations as $category => $messages) {
    foreach ($messages as $message => $translation) {
        $command->bindValues([':category' => $category, ':message'  => $message]);
        $command->execute(); // << Warning goes here
    }
}
```
Writing inner foreach as `foreach ($messages as $message => $translation) {` doesn't change anything.

**Comment by kalessil** [comment]
@TrueXakeP: same as in #1431 - the inspection is not recognizing objects state chage yet. Please follow #1431 to stay up=todate with updates.
